### DISCUSSION TITLE
offset commits on Async tasks

### DISCUSSION THREAD

**Discussion by duke-bartholomew** [LGouellec/streamiz#243] [Q&A] [answered]
Hi, I have a question regarding the async (sink)operations in kafka-streams-dotnet.
Assume I have following topology:

```csharp
    var builder = new StreamBuilder();
    builder.Stream<string, string>("my-topic")
        .MapValues(JsonConvert.DeserializeObject<MyObject>)
        .ForeachAsync(async (record, _) => {
                db.MyObjects.Add(record.Value);
                await db.SaveChangesAsync();
            },
            RetryPolicy.NewBuilder()
                .NumberOfRetry(10)
                .RetryBackOffMs(100)
                .RetriableException<DatabaseException>()
                .RetryBehavior(EndRetryBehavior.FAIL)
                .Build());
```
.. so writing to a database asynchronously.

**What happens under the hood in kafka-streams-dotnet**?
I read that an "external streamthread" will be used to handle that action.  
Does this mean a new thread per message in this case? (I come from the JVM world, so my perception of threads and the impact of them might be skewed when it comes to .NET threads)

**What about offset commits**?
Assume 5 messages are in-flight of being inserted in the database (given above example topology).  What would the committed offset be in case message 2,3,4 and 5 are properly inserted, but message 1 fails after it's retry period?

Also, assume I have successfully processed messages 1 to 5 and have message 6 to 20 in-flight in the async block.
What offset will be committed in this case?

And a related question:
`threading-model.rst` states the following:
> For each async processor, a request and a response (except for ForEachAsync(..)) topics will be created and the stream thread publish into the request topic and subscribe the response topic.

So this would mean that using `ForEachAsync` will not create such request/response topics?
(but while doing some quick tests, these topics do pop up on my Kafka cluster)

**Comment by LGouellec** [comment, marked as answer]
Hi @duke-bartholomew , 

**What happens under the hood in kafka-streams-dotnet?**
_Does this mean a new thread per message in this case? (I come from the JVM world, so my perception of threads and the impact of them might be skewed when it comes to .NET threads)_

No a single thread is created and dedicated to consume the internal request, call the external processor with logic retry and so on, and publish the result in the internal response topic. This thread has a logic with pause/resume in case the processor is too long compared to `max.poll.interval.ms`. Btw, in JAVA the consumer has `max.poll.interval.ms` to consume a batch records between 0 and `max.poll.records` messages whereas in .Net the consumer has `max.poll.interval.ms` to consume one message. It's quite different. 
All the logic is present [here](https://github.com/LGouellec/kafka-streams-dotnet/blob/b700b698bb45dee5f51a49876d64c59b17432399/core/Processors/ExternalStreamThread.cs#L131), [here](https://github.com/LGouellec/kafka-streams-dotnet/blob/b700b698bb45dee5f51a49876d64c59b17432399/core/Processors/ExternalProcessorTopologyExecutor.cs#L160).

**What about offset commits?**
_Assume 5 messages are in-flight of being inserted in the database (given above example topology). What would the committed offset be in case message 2,3,4 and 5 are properly inserted, but message 1 fails after it's retry period?

Also, assume I have successfully processed messages 1 to 5 and have message 6 to 20 in-flight in the async block.
What offset will be committed in this case?_

The external DSL allow async/await syntax but some reasons like you mentioned, I can't execute the processing fully asynchronously. 
Because if the first request throw a timed out and the message is not inserted whereas next message do then I can commit message and "loose" record 1.

The commit is called every `commit.interval.ms` like the other stream thread after the processing. So for each async/await call, I wait the task is done [here](https://github.com/LGouellec/kafka-streams-dotnet/blob/b700b698bb45dee5f51a49876d64c59b17432399/core/Processors/AbstractAsyncProcessor.cs#LL55C31-L55C31). 

A new feature could be execute all the batch of records fetched from the consumer asynchronously and waiting all the task before committing the offset and fetch the next batch of record. I keep in mind.

_So this would mean that using ForEachAsync will not create such request/response topics?
(but while doing some quick tests, these topics do pop up on my Kafka cluster)_

When you use `ForEachAsync(...)`, there is only a request topic. Because you don't have a new stream returning the foreach processor. So you have a internal request topic consumed by the external thread, but no output endpoint.

Hope it's clear

Best regards,

**Reply by duke-bartholomew** [reply]
Hi @LGouellec 

Aha, so it's basically a 2-stepped approach then?

```
[ source (TOPIC: my-input-stream) ] ---> [ process ] ---> [ ForEachAsync: (TOPIC:  process-000-request-sink)]
```
and
```
[ source (TOPIC: process-000-request-sink) ] ---> [ doAction ]
                                                     +=> with 'doAction' working it's way through 
                                                         the messages one-by-one (synchronous)
```
With offset commits happening automatically (on interval) on the `my-input-stream` partitions (after they got produced to the `request-sink` topic) and offset commits on the `process-000-request-sink` partitions when successfully processed a message.

**Reply by LGouellec** [reply]
Yes this exactly this king of behavior

**Comment by mayur-rao** [comment]
Hi @LGouellec, and thanks for this library — and for your detailed answer in #243, which is what prompted this question.

## Setup

We use `ForeachAsync` as a terminal processor. To reduce load on a downstream **external, idempotent RPC service**, we don't write per-record. Instead the delegate **buffers** each record in an in-memory structure and only **flushes in bulk** (on a size threshold or a timer) to that service:

```csharp
.ForeachAsync(async (record, _) =>
{
    _buffer.Add(Map(record));           // ~most records: just buffer, then return
    if (SizeReached() || TimerExpired())
        await SendBulkToExternalService(_buffer.DrainSnapshot());  // ~1 record in N
});
```

## What we understand from #243

A single external thread runs the delegate sequentially, and offsets are committed every `commit.interval.ms` **after the delegate returns**. There's no per-record commit and no built-in "don't commit this one" hook.

## The problem

Because most records return **immediately after buffering** (no external write yet), their offsets become eligible to commit **before** the bulk flush that actually persists them. If the process crashes between "offset committed" and "bulk flush durably written," those records are **silently lost** — Kafka thinks they're done, but they never reached the sink.

The sink is an external RPC (can't enlist in a Kafka transaction), and our targets are idempotent, so we're aiming for **effectively-once / at-least-once with no data loss**, not strict EOS.

## Questions

1. Is our understanding correct that with `ForeachAsync` there's **no supported way to withhold/defer** an offset commit for records that returned from the delegate but are still pending in an app-level buffer?
2. Is there a recommended way to **own or defer commits** in this scenario — e.g. disabling auto-commit, a manual commit API, `ProcessorContext.Commit()`, or dropping to the **Processor API** with a custom `IProcessor` that commits only after the async bulk send succeeds?
3. In #243 you mentioned a possible future feature: *fetch a batch of records, await all async tasks, then commit and fetch the next batch.* Did anything like that land in a later release? We're on **1.5.x**; latest is **1.8.x**.
4. For a "batch-to-external-sink" pattern like this, what would you recommend to guarantee no data loss? Is the intended approach essentially to **flush synchronously within the delegate** (so return ≈ durable), accepting the throughput cost — or is there a cleaner idiom we're missing?
5. If we do own commits manually, are there interactions we should watch for with the **BUFFERED retry policy** (pause/resume, in-memory retry buffer)?

Thanks in advance

**Comment by LGouellec** [comment]
Hi @mayur-rao , 

Sorry for the delay, 

> 1- Is our understanding correct that with ForeachAsync there's no supported way to withhold/defer an offset commit for records that returned from the delegate but are still pending in an app-level buffer?

Correct, even for the classic API, when the message leave the processing topology, it's considering as processed and acked. So even if you implement the same thing with a classic Peek or ForEach, you'll face the same behavior. Same in KStreams JAVA btw.


 > 2 - Is there a recommended way to own or defer commits in this scenario — e.g. disabling auto-commit, a manual commit API, ProcessorContext.Commit(), or dropping to the Processor API with a custom IProcessor that commits only after the async bulk send succeeds?
 
 No unfortunately , there is no workaround. Except persisting your buffer with a rocksdb instance backed with a changelog topic
 
 > 3 - In https://github.com/LGouellec/streamiz/discussions/243 you mentioned a possible future feature: fetch a batch of records, await all async tasks, then commit and fetch the next batch. Did anything like that land in a later release? We're on 1.5.x; latest is 1.8.x.
 
 I plan to refactor all the external processing topology in 1.9.X. See [Issue 360](https://github.com/LGouellec/streamiz/issues/360), but the development is lot of delayed. 
 
 > 4 - For a "batch-to-external-sink" pattern like this, what would you recommend to guarantee no data loss? Is the intended approach essentially to flush synchronously within the delegate (so return ≈ durable), accepting the throughput cost — or is there a cleaner idiom we're missing?
 
 I would use a local state store backed with a Kafka topic, and use a Punctuator to sink every X seconds, and manage the state (meaning the ack) manually. So even if the application failed, the messages are already committed, but in your state, the buffer is still there. 
 
 > If we do own commits manually, are there interactions we should watch for with the BUFFERED retry policy (pause/resume, in-memory retry buffer)?
 
 Pause and Resume can lead processing twice the same messages with Librdkafka.
